internal/raft/message/convert.go

```
@@ -0,0 +1,971 @@
+	msgExprFunc := &Expr_Func{&FunctionExpr{}}
+	msgExprFunc.Func.Name = cmd.Name
+	msgExprFunc.Func.Distinct = cmd.Distinct
+	msgExprFunc.Func.Args = ConvertCommandToMessageRepeatedExpr(cmd.Args)
+	return msgExprFunc
+}
+
+// ConvertCommandToMessageEqualityExpr converts a command.Expr to a message.Expr_Equality
+func ConvertCommandToMessageEqualityExpr(cmd *command.EqualityExpr) *Expr_Equality {
+	msgExprEquality := &Expr_Equality{&EqualityExpr{}}
+	msgExprEquality.Equality.Left = ConvertCommandToMessageExpr(cmd.Left)
+	msgExprEquality.Equality.Right = ConvertCommandToMessageExpr(cmd.Right)
+	msgExprEquality.Equality.Invert = cmd.Invert
+	return msgExprEquality
+}
+
+// ConvertCommandToMessageRangeExpr converts a command.Expr to a message.Expr_Range
+func ConvertCommandToMessageRangeExpr(cmd *command.RangeExpr) *Expr_Range {
+	msgExprRange := &Expr_Range{&RangeExpr{}}
+	msgExprRange.Range.Needle = ConvertCommandToMessageExpr(cmd.Needle)
+	msgExprRange.Range.Lo = ConvertCommandToMessageExpr(cmd.Lo)
+	msgExprRange.Range.Hi = ConvertCommandToMessageExpr(cmd.Hi)
+	msgExprRange.Range.Invert = cmd.Invert
+	return msgExprRange
+}
+
+// ConvertCommandToMessageExpr converts command.Expr to a message.Expr
+func ConvertCommandToMessageExpr(cmd command.Expr) *Expr {
+	msgExpr := &Expr{}
+	switch cmd := cmd.(type) {
+	case *command.LiteralExpr:
+		msgExpr.Expr = ConvertCommandToMessageLiteralExpr(cmd)
+	case *command.ConstantBooleanExpr:
+		msgExpr.Expr = ConvertCommandToMessageConstantBooleanExpr(cmd)
+	case *command.UnaryExpr:
+		msgExpr.Expr = ConvertCommandToMessageUnaryExpr(cmd)
+	case *command.BinaryExpr:
+		msgExpr.Expr = ConvertCommandToMessageBinaryExpr(cmd)
+	case *command.FunctionExpr:
+		msgExpr.Expr = ConvertCommandToMessageFunctionalExpr(cmd)
+	case *command.EqualityExpr:
+		msgExpr.Expr = ConvertCommandToMessageEqualityExpr(cmd)
+	case *command.RangeExpr:
+		msgExpr.Expr = ConvertCommandToMessageRangeExpr(cmd)
+	}
```

No default case, what happens if there's another expression that is passed in? Probably should return an error in that case.

---

How would error be returned in type Expr?

---

An additional return argument